File: web/src/services/release.ts
import request from './api'
import type { ApiResponse, PageResponse } from '../types'

type RequestPromise<T> = Promise<T>

export interface Release {
  id?: number
  created_at?: string
  updated_at?: string
  title: string
  application_id?: number
  application_name: string
  project_id?: number
  project_name?: string
  env: string
  version: string
  description: string
  status: string
  risk_level: string
  created_by: number
  created_by_name: string
  approved_by?: number
  approved_by_name: string
  approved_at?: string
  published_at?: string
  published_by?: number
  published_by_name: string
  rollback_at?: string
  reject_reason: string
  pipeline_runs?: any[]
  nacos_releases?: any[]
  biz_version_id?: number
  biz_version_name?: string
  biz_goal_id?: number
  biz_goal_name?: string
  // ---------- v2.0 字段 ----------
  rollout_strategy?: 'direct' | 'canary' | 'blue_green'
  rollout_config?: Record<string, any>
  risk_score?: number
  risk_factors?: {
    score: number
    level: string
    hits: Array<{ key: string; name: string; weight: number; detail?: string }>
    calculated: string
    version: string
  }
  approval_instance_id?: number
  gitops_change_request_id?: number
  argo_app_name?: string
  argo_sync_status?: string
  jira_issue_keys?: string
}

// v2.0: GitOps PR 请求 / 响应
export interface GitOpsPRRequest {
  target_branch?: string
  commit_message?: string
  dry_run?: boolean
}

export interface GitOpsPRResponse {
  change_request_id: number
  pr_url?: string
  branch_name: string
  files_changed: string[]
  dry_run: boolean
  message?: string
}

export interface ReleaseItem {
  id?: number
  created_at?: string
  release_id: number
  item_type: string
  item_id: number
  item_title: string
  item_status: string
  sort_order: number
}

export interface ReleaseFilter {
  env?: string
  status?: string
  application_id?: number
  project_id?: number
  title?: string
  page?: number
  page_size?: number
}

export interface CreateReleaseFromPipelineRunRequest {
  pipeline_run_id: number
  existing_release_id?: number
  title?: string
  env?: string
  version?: string
  description?: string
  risk_level?: string
  rollout_strategy?: 'direct' | 'canary' | 'blue_green'
  rollout_config?: Record<string, any>
}

export interface ReleaseOverviewStage {
  key: string
  label: string
  status: 'wait' | 'process' | 'finish' | 'error' | string
  message?: string
}

export interface ReleaseOverview {
  release_id: number
  status: string
  current_stage: string
  blocked: boolean
  blocked_reason?: string
  next_action: string
  approval: {
    instance_id?: number
    status: string
    chain_name?: string
    current_node_order?: number
    started_at?: string
    finished_at?: string
  }
  gitops: {
    change_request_id?: number
    status: string
    mr_url?: string
    approval_status?: string
    auto_merge_status?: string
    error_message?: string
    updated_at?: string
  }
  argocd: {
    application_id?: number
    app_name?: string
    sync_status?: string
    health_status?: string
    drift_detected: boolean
    last_sync_at?: string
  }
  stages: ReleaseOverviewStage[]
}

export interface ReleaseGateResult {
  key: string
  name: string
  category: string
  status: 'pass' | 'warn' | 'block' | 'skip' | string
  severity: string
  policy: string
  blocker: boolean
  message: string
  detail?: Record<string, any>
  evaluated_at: string
}

export interface ReleaseGateSummary {
  release_id: number
  status: 'pass' | 'warn' | 'block' | string
  blocked: boolean
  can_publish: boolean
  block_reasons: string[]
  warn_reasons: string[]
  next_action: string
  evaluated_at: string
  items: ReleaseGateResult[]
}

export const releaseApi = {
  list(params: ReleaseFilter) {
    return request.get('/releases', { params }) as RequestPromise<ApiResponse<PageResponse<Release>>>
  },
  getById(id: number) {
    return request.get(`/releases/${id}`) as RequestPromise<ApiResponse<Release>>
  },
  getOverview(id: number) {
    return request.get(`/releases/${id}/overview`) as RequestPromise<ApiResponse<ReleaseOverview>>
  },
  getGates(id: number) {
    return request.get(`/releases/${id}/gates`) as RequestPromise<ApiResponse<ReleaseGateSummary>>
  },
  refreshGates(id: number) {
    return request.post(`/releases/${id}/gates/refresh`) as RequestPromise<ApiResponse<ReleaseGateSummary>>
  },
  create(data: Partial<Release>) {
    return request.post('/releases', data) as RequestPromise<ApiResponse<Release>>
  },
  createFromPipelineRun(data: CreateReleaseFromPipelineRunRequest) {
    return request.post(`/delivery/pipeline-runs/${data.pipeline_run_id}/release`, data) as RequestPromise<ApiResponse<Release>>
  },
  update(id: number, data: Partial<Release>) {
    return request.put(`/releases/${id}`, data) as RequestPromise<ApiResponse<Release>>
  },
  delete(id: number) {
    return request.delete(`/releases/${id}`) as RequestPromise<ApiResponse<void>>
  },
  listItems(id: number) {
    return request.get(`/releases/${id}/items`) as RequestPromise<ApiResponse<ReleaseItem[]>>
  },
  addItem(id: number, data: { item_type: string; item_id: number; item_title?: string }) {
    return request.post(`/releases/${id}/items`, data) as RequestPromise<ApiResponse<void>>
  },
  removeItem(releaseId: number, itemId: number) {
    return request.delete(`/releases/${releaseId}/items/${itemId}`) as RequestPromise<ApiResponse<void>>
  },
  submit(id: number) {
    return request.post(`/releases/${id}/submit`) as RequestPromise<ApiResponse<Release>>
  },
  approve(id: number) {
    return request.post(`/releases/${id}/approve`) as RequestPromise<ApiResponse<Release>>
  },
  reject(id: number, reason: string) {
    return request.post(`/releases/${id}/reject`, { reason }) as RequestPromise<ApiResponse<Release>>
  },
  publish(id: number) {
    return request.post(`/releases/${id}/publish`) as RequestPromise<ApiResponse<Release>>
  },

  // ---------- v2.0 ----------
  /** 真实触发 GitOps PR 生成 */
  openGitOpsPR(id: number, body: GitOpsPRRequest = {}) {
    return request.post(`/releases/${id}/gitops-pr`, body) as RequestPromise<ApiResponse<GitOpsPRResponse>>
  },
  /** dry-run 预览 GitOps PR 将影响的文件列表（不调用 ArgoCD/Git） */
  dryRunGitOpsPR(id: number, body: GitOpsPRRequest = {}) {
    return request.post(`/releases/${id}/gitops-pr/dry-run`, body) as RequestPromise<ApiResponse<GitOpsPRResponse>>
  },
}

File: pkg/dto/release.go
// Package dto
//
// release.go: 发布主单（Release）相关的请求/响应 DTO（v2.0）。
// 配合 internal/models/deploy/release.go 使用。
package dto

import "time"

// ReleaseV2DTO 发布主单的列表/详情统一响应体（v2.0）。
//
// 注：用 V2 后缀以与未来可能引入的 ReleaseV1DTO 区分；
// 字段命名与 model.Release 保持一致以减少前后端心智负担。
type ReleaseV2DTO struct {
	ID              uint       `json:"id"`
	Title           string     `json:"title"`
	ApplicationID   *uint      `json:"application_id"`
	ApplicationName string     `json:"application_name"`
	Env             string     `json:"env"`
	Version         string     `json:"version"`
	Description     string     `json:"description"`
	Status          string     `json:"status"`
	RiskLevel       string     `json:"risk_level"`
	RiskScore       int        `json:"risk_score"`
	RolloutStrategy string     `json:"rollout_strategy"`
	CreatedBy       uint       `json:"created_by"`
	CreatedByName   string     `json:"created_by_name"`
	CreatedAt       time.Time  `json:"created_at"`
	ApprovedAt      *time.Time `json:"approved_at,omitempty"`
	ApprovedByName  string     `json:"approved_by_name,omitempty"`
	PublishedAt     *time.Time `json:"published_at,omitempty"`
	PublishedByName string     `json:"published_by_name,omitempty"`

	// v2.0 新增字段
	ApprovalInstanceID    *uint                  `json:"approval_instance_id,omitempty"`
	GitOpsChangeRequestID *uint                  `json:"gitops_change_request_id,omitempty"`
	ArgoAppName           string                 `json:"argo_app_name,omitempty"`
	ArgoSyncStatus        string                 `json:"argo_sync_status,omitempty"`
	JiraIssueKeys         []string               `json:"jira_issue_keys,omitempty"`
	RolloutConfig         map[string]interface{} `json:"rollout_config,omitempty"`
	RiskFactors           map[string]interface{} `json:"risk_factors,omitempty"`

	// 子项汇总（按类型计数，UI 用于"3 部署 / 1 配置 / 2 数据库" 这种概览）
	ItemSummary map[string]int `json:"item_summary,omitempty"`
}

// CreateReleaseRequest 创建发布主单请求体。
type CreateReleaseRequest struct {
	Title           string                 `json:"title" binding:"required,max=200"`
	ApplicationID   *uint                  `json:"application_id"`
	ApplicationName string                 `json:"application_name"`
	Env             string                 `json:"env" binding:"required"`
	Version         string                 `json:"version"`
	Description     string                 `json:"description"`
	RiskLevel       string                 `json:"risk_level"`
	RolloutStrategy string                 `json:"rollout_strategy"` // direct/canary/blue_green
	RolloutConfig   map[string]interface{} `json:"rollout_config,omitempty"`
	JiraIssueKeys   []string               `json:"jira_issue_keys,omitempty"`
}

// CreateReleaseFromPipelineRunRequest 从一次成功 CI 运行生成发布主单。
// 这是应用交付主链路中 CI -> Release 的标准入口。
type CreateReleaseFromPipelineRunRequest struct {
	PipelineRunID     uint                   `json:"pipeline_run_id" binding:"required"`
	ExistingReleaseID uint                   `json:"existing_release_id,omitempty"`
	Title             string                 `json:"title"`
	Env               string                 `json:"env"`
	Version           string                 `json:"version"`
	Description       string                 `json:"description"`
	RiskLevel         string                 `json:"risk_level"`
	RolloutStrategy   string                 `json:"rollout_strategy"`
	RolloutConfig     map[string]interface{} `json:"rollout_config,omitempty"`
}

// UpdateReleaseRequest 编辑发布主单请求体（仅 draft 状态可编辑）。
type UpdateReleaseRequest struct {
	Title           string                 `json:"title" binding:"max=200"`
	Description     string                 `json:"description"`
	RiskLevel       string                 `json:"risk_level"`
	RolloutStrategy string                 `json:"rollout_strategy"`
	RolloutConfig   map[string]interface{} `json:"rollout_config,omitempty"`
	JiraIssueKeys   []string               `json:"jira_issue_keys,omitempty"`
}

// ReleaseListResponse 列表响应。
type ReleaseListResponse struct {
	List       []ReleaseV2DTO `json:"list"`
	Total      int64          `json:"total"`
	Page       int            `json:"page"`
	PageSize   int            `json:"pageSize"`
	TotalPages int            `json:"totalPages"`
}

// ReleaseOverviewDTO 聚合发布主链路状态，供 Release 详情首屏使用。
type ReleaseOverviewDTO struct {
	ReleaseID     uint                      `json:"release_id"`
	Status        string                    `json:"status"`
	CurrentStage  string                    `json:"current_stage"`
	Blocked       bool                      `json:"blocked"`
	BlockedReason string                    `json:"blocked_reason,omitempty"`
	NextAction    string                    `json:"next_action"`
	Approval      ReleaseOverviewApproval   `json:"approval"`
	GitOps        ReleaseOverviewGitOps     `json:"gitops"`
	ArgoCD        ReleaseOverviewArgoCD     `json:"argocd"`
	Stages        []ReleaseOverviewStageDTO `json:"stages"`
}

type ReleaseOverviewStageDTO struct {
	Key     string `json:"key"`
	Label   string `json:"label"`
	Status  string `json:"status"` // wait/process/finish/error
	Message string `json:"message,omitempty"`
}

type ReleaseOverviewApproval struct {
	InstanceID       *uint      `json:"instance_id,omitempty"`
	Status           string     `json:"status"`
	ChainName        string     `json:"chain_name,omitempty"`
	CurrentNodeOrder int        `json:"current_node_order,omitempty"`
	StartedAt        *time.Time `json:"started_at,omitempty"`
	FinishedAt       *time.Time `json:"finished_at,omitempty"`
}

type ReleaseOverviewGitOps struct {
	ChangeRequestID *uint      `json:"change_request_id,omitempty"`
	Status          string     `json:"status"`
	MRURL           string     `json:"mr_url,omitempty"`
	ApprovalStatus  string     `json:"approval_status,omitempty"`
	AutoMergeStatus string     `json:"auto_merge_status,omitempty"`
	ErrorMessage     string     `json:"error_message,omitempty"`
	UpdatedAt        *time.Time `json:"updated_at,omitempty"`
}

type ReleaseOverviewArgoCD struct {
	ApplicationID *uint      `json:"application_id,omitempty"`
	AppName       string     `json:"app_name,omitempty"`
	SyncStatus    string     `json:"sync_status,omitempty"`
	HealthStatus  string     `json:"health_status,omitempty"`
	DriftDetected bool       `json:"drift_detected"`
	LastSyncAt     *time.Time `json:"last_sync_at,omitempty"`
}

// GitOpsPRRequest 触发 GitOps PR 生成的请求体（v2.0）。
//
// 调用前置：Release 已 approved。
type GitOpsPRRequest struct {
	ReleaseID    uint   `json:"release_id" binding:"required"`
	TargetBranch string `json:"target_branch"` // 默认 main
	CommitMsg    string `json:"commit_message,omitempty"`
	DryRun       bool   `json:"dry_run,omitempty"`
}

// GitOpsPRResponse GitOps PR 生成结果。
type GitOpsPRResponse struct {
	ChangeRequestID uint     `json:"change_request_id"`
	PRURL           string   `json:"pr_url,omitempty"`
	BranchName      string   `json:"branch_name"`
	FilesChanged    []string `json:"files_changed"`
	DryRun          bool     `json:"dry_run"`
	Message         string   `json:"message,omitempty"`
}
